### PR TITLE
don't save Modules in hparams

### PR DESCRIPTION
**Pull request by KnathanM**
## Description
Related to #898, this PR address the fact that we double save `nn.Modules` in both the `hparams` and in the `state_dict`. This PR follows David's idea [here](https://github.com/chemprop/chemprop/pull/898#issuecomment-2158733975). 

The main downside to this PR is a lot more code is required to implement this versus continuing to double save. The main upside is that model files will be a little smaller. 

## Other notes
`metrics` in `MPNN` is a list of `Modules` so the whole list is pickled and unpickled to save and load. This is different than all the other Modules like `criterion` which are rebuilt each time a model is loaded. 

The default `task_weights` for `criterion` is an array, so I don't need to check if there are `task_weights` in the `state_dict`. For the other cases, the default is `None`/`Identity` so in those cases, I need to check if their are corresponding parameters in the `state_dict` and if so build a `Module` with zeros. 

While working on this PR, I found that for multicomponent message passing, if a single block is shared between two components, the weights for that block appear twice in the `state_dict`. I don't know if the weights just have two key values, or if they are truly saved twice in the model file. Not sure if this is something we should try to change.

### CODE REVIEW

File: chemprop/models/model.py [hunk 1]
```diff
@@ -228,41 +228,84 @@ def configure_optimizers(self):
...
+            hparams = d["hyper_parameters"]
+            state_dict = d["state_dict"]
+        except KeyError:
+            raise KeyError(f"Could not find hyper parameters and/or state dict in {path}.")
+
+        if "message_passing.V_d_transform.mean" in state_dict:
+            dummy_tensor = (
+                torch.zeros_like(state_dict["message_passing.V_d_transform.mean"])
+                .squeeze(0)
+                .numpy()
```
Comment (KnathanM) [inline chemprop/models/model.py:244]: `_ScaleTransformMixin` unsqueezes and converts to a torch tensor. If I don't include `.numpy()` here, I get this warning: `UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  scale = torch.cat([torch.ones(pad), torch.tensor(scale, dtype=torch.float)])`

**Comment by davidegraff** [comment]
woof- on second thought, i don't think my idea was a good one. I don't think it's that much of a problem to double-save like 32 floating point numbers

**Comment by KnathanM** [comment]
Okay, sounds like we will go with #898 for saving the scalers. 

Questions:
I added some unit tests here. Are those worth still merging in? They test if the scalers are saved and loaded correctly, which is independent of the mechanics how we save and load them. 

I reorganized some of `load_from_checkpoint` and `load_from_file` by having the hyperparameters and state dict loaded in `load_submodules`. This allowed removing a separate `load_from_file` in `MulticomponentMPNN`. Is this worth merging in?

**Comment by KnathanM** [comment]
I moved the tests to #955 

I decided to not move the changes to `load_from_file` to another PR because I think having `MPNN` and `MulticomponentMPNN` each have their own `load_from_file` may help people debug in the future.
